models/context_branch.py

```diff
@@ -0,0 +1,296 @@
+        src = src + self.dropout3(src2)
+        src = self.norm2(src)
+        return src
+
+    def forward(self, src, pos, reference_points, spatial_shapes, level_start_index):
+        # self attention
+        src2 = self.self_attn(self.with_pos_embed(src, pos), reference_points, src, spatial_shapes, level_start_index)
+        src = src + self.dropout1(src2)
+        src = self.norm1(src)
+
+        # ffn
+        src = self.forward_ffn(src)
+
+        return src
+
+class Encoder(nn.Module):
+    def __init__(self, encoder_layer, num_layers):
+        super().__init__()
+        self.layers = _get_clones(encoder_layer, num_layers)
+        self.num_layers = num_layers
+
+    @staticmethod
+    def get_reference_points(spatial_shapes, device):
+        
+        reference_points_list = []
+        for lvl, (H_, W_) in enumerate(spatial_shapes):
+
+            ref_y, ref_x = torch.meshgrid(torch.linspace(0.5, H_ - 0.5, H_, dtype=torch.float32, device=device),
+                                          torch.linspace(0.5, W_ - 0.5, W_, dtype=torch.float32, device=device))
+            ref_y = ref_y.reshape(-1)[None] / H_
+            ref_x = ref_x.reshape(-1)[None] / W_
+            ref = torch.stack((ref_x, ref_y), -1)
+            reference_points_list.append(ref)
+        
+        reference_points = torch.cat(reference_points_list, 1)
+        
+        reference_points = reference_points[:, :, None].repeat(1, 1, len(spatial_shapes), 1)
+
+        return reference_points
+
+    def forward(self, src, spatial_shapes, level_start_index, pos):
+        output = src
+
+        reference_points = self.get_reference_points(spatial_shapes, device=src.device)
+
+        for _, layer in enumerate(self.layers):
+            output = layer(output, pos, reference_points, spatial_shapes, level_start_index)
+        
+        return output
+
+class DecoderLayer(nn.Module):
+    def __init__(self, d_model, d_ffn, dropout, activation, n_levels, n_heads, n_points):
+        
+        super().__init__()
+
+        # cross attention
+        self.cross_attn = MSDeformAttn(d_model, n_levels, n_heads, n_points)
+        self.dropout1 = nn.Dropout(dropout)
+        self.norm1 = nn.LayerNorm(d_model)
+
+        # self attention
+        self.self_attn = nn.MultiheadAttention(d_model, n_heads, dropout=dropout)
+        self.dropout2 = nn.Dropout(dropout)
+        self.norm2 = nn.LayerNorm(d_model)
+
+        # ffn
+        self.linear1 = nn.Linear(d_model, d_ffn)
+        self.activation = _get_activation_fn(activation)
+        self.dropout3 = nn.Dropout(dropout)
+        self.linear2 = nn.Linear(d_ffn, d_model)
+        self.dropout4 = nn.Dropout(dropout)
+        self.norm3 = nn.LayerNorm(d_model)
+    
+    @staticmethod
+    def with_pos_embed(tensor, pos):
+        return tensor if pos is None else tensor + pos
+
+    @staticmethod
+    def get_reference_points(spatial_shapes, device, h, w):
+        
+        # H, W = spatial_shapes[0]
```

Comment:
I think here is hard-coded, is there a way to make this adaptable? H and W should be passed from outside, or at least defined during model initialization.